### PR TITLE
feat(guard): run scoped step-safety inference with ONNX

### PR DESCRIPTION
**Pull request by hasandemirkiran**
## Summary

Run the local step-safety pilot from Go through CPU ONNX Runtime, with no Python dependency on endpoints. The model remains disabled by default, advisory-only, fail-open, and independent of Kestrel and authoritative policy.

- Exclude Read/Write/Edit-style tools from both scoring and history. Preserve complete requests, candidate actions, and schemas; skip them when they exceed their budgets.
- Keep the 512-token training layout: request 96, history 144, action 128, schemas 128. After file-tool exclusions, truncate history to its first 72 and last 72 tokens, matching training. Supported results above 4 KiB now reach token packing; the history cache remains bounded to 64 KiB and 24 events per session.
- Use pure-Go tokenization/packing, verified artifacts, and an explicit native-runtime installer. Preserve CGO-free builds for macOS/Linux on amd64/arm64.
- Bound outstanding inference to one call. Return at the hook deadline, discard late native results, and retain the slot until native work finishes. Exclusions, omissions, and failures are recorded without fabricating a safe score.
- Resolve the merge conflicts and preserve current hook metadata and socket/restart behavior. Version the runtime/input policy as `toolsafe-deberta-v3-xsmall-onnx-scoped-v2`; checkpoint weights, calibration, and threshold are unchanged.

## Coverage findings

The checkpoint was trained on 2,192 TS-Bench rows: 1,518 ASB and 674 AgentAlign. It provides a narrow advisory signal; direct training coverage does not establish support for Read/Write/Edit tools. Kestrel's shell signal remains separate.

Replaying the same archived inputs before and after history truncation gives the following results for the panel selected solely for input eligibility:

| Source | Unsafe detected before → after | Safe flagged before → after |
| --- | --- | --- |
| ASB | 20/20 → 20/20 | 0/20 → 0/20 |
| AgentAlign validation | 5/5 → 5/5 | 0/12 → 0/12 |
| AgentDojo | 0/4 → 3/4 | 0/2 → 0/2 |
| AgentHarm | 1/3 → 1/3 | 0/8 → 2/8 |

The recovered detections and additional false alarms match the training-reference behavior. None requires restoring Read. These are reused benchmark/validation controls, not a new independent holdout. Admission is still restrictive outside ASB: a separate balanced 40-example panel per source scores 40 ASB, 3 AgentAlign, 0 AgentDojo, and 3 AgentHarm examples, mostly because archived schemas are too large. Live adapters may supply different schemas, so those are not production coverage estimates.

The original 48-case synthetic smoke test still skips 10 file-tool cases and scores 38, with **0/19 unsafe examples detected** and no false alarms among 19 safe examples. Many examples supplied their context through the excluded Read history. Software correctness should not be confused with broad model usefulness. These findings support an opt-in shadow pilot to measure coverage and errors, not enforcement.

Sampling, the 192 unique case IDs, and before/after counts are recorded in `docs/step-safety-shadow-results.json`.

## Validation

- `go build ./...`, `go vet ./...`, `go test -race ./...`, and binary `govulncheck` passed with Go 1.25.13.
- CGO-free release builds passed for macOS/Linux on amd64/arm64.
- 532 Hugging Face tokenizer vectors matched exactly; seven frozen checkpoint cases matched packed tokens exactly and logits within `2e-5`.
- Native ONNX parity, timeout/next-request, and large-history cache-to-Evaluator tests passed on Apple Silicon and Linux ARM64. Large-history fixtures match independently generated training token vectors and reference scores.
- The aligned replay reused the same 192 unique inputs and verified reference token vectors, with no inference failures or timeouts.
- Guard smoke and real-model hook-to-SQLite E2E passed: Bash scored, Read excluded, a large supported tool result reached training-aligned inference, and authorization stayed unchanged.
- Verified installation and reproducible ONNX export are documented alongside the existing local CPU latency measurements in `docs/step-safety-shadow.md`.

Python is used only for maintainer export/reference tooling; serving and ordinary Go tests do not require it.

### CODE REVIEW

**Comment by hasandemirkiran** [comment]
* **#500** <a href="https://app.graphite.com/github/pr/kontext-security/kontext/500?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#478** <a href="https://app.graphite.com/github/pr/kontext-security/kontext/478?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/kontext-security/kontext/478?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>queue_me</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/kontext-security?ref=merge-queue-instructions-in-stack-comment&prId=478).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: internal/guard/app/server/runtime.go [hunk 1]
```diff
@@ -126,10 +137,14 @@ func (r guardHookRuntime) decideAndRecord(ctx context.Context, event risk.HookEv
...
 	}
 	r.observePrompt(event)
+	r.observeStepContext(event)
 	decision, err := r.policy.DecideHook(ctx, event)
 	if err != nil {
 		return risk.RiskDecision{}, err
 	}
+	// This synchronous shadow call is the last model stage before a successful
+	// PreToolUse response can release the tool. It never mutates decision.
+	r.annotateStepSafety(ctx, event, &decision)
```
[the remark this answers is not in the document]
Comment (hasandemirkiran) [reply internal/guard/app/server/runtime.go:156]: Fixed in 05c7be3. Managed PreToolUse now has a one-second edge budget, separate from the original policy allowance; the shadow timeout remains 250 ms by default and is capped at 500 ms. Timed-out workers are killed immediately and reaped off-path. TestLifecycleReservesShadowBudgetAfterAuthoritativePolicy proves an authoritative enforce allow still returns after a response longer than the former complete 250 ms window.

File: internal/guard/app/server/runtime.go [hunk 2]
```diff
@@ -172,9 +188,91 @@ func (r guardHookRuntime) decideAndRecord(ctx context.Context, event risk.HookEv
...
+	case hook.HookUserPromptSubmit.String():
+		prompt, _ := event.ToolInput["prompt"].(string)
+		r.stepContext.RecordUserRequest(event.SessionID, prompt)
+	case hook.HookPostToolUse.String(), hook.HookPostToolUseFailed.String():
+		r.stepContext.RecordInteraction(event.SessionID, stepsafety.HistoryEntry{
+			ToolName:      event.ToolName,
+			ToolArguments: event.ToolInput,
+			ToolResponse:  event.ToolResponse,
+			Error:         event.Error,
+		})
```
[the remark this answers is not in the document]
Comment (hasandemirkiran) [reply internal/guard/app/server/runtime.go:252]: Fixed in 05c7be3. RuntimeCore now exposes a memory-only AsyncEventObserver. localruntime records structured PostToolUse context synchronously before acknowledging it, then calls IngestObservedEvent for durable background work so the context is not applied twice. Socket-level and service tests verify the immediately following PreToolUse sees the prior interaction exactly once.

File: internal/guard/stepsafety/worker.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    input_ids = [int(start_id)]
+    for name in ("request", "history", "action", "schema"):
+        marker_id = tokenizer.convert_tokens_to_ids(FIELD_MARKERS[name])
+        if marker_id is None or marker_id == tokenizer.unk_token_id:
+            raise ValueError(f"Tokenizer is missing field marker {FIELD_MARKERS[name]}")
+        tokens = tokenizer.encode(fields[name], add_special_tokens=False)
+        selected = (
+            _head(tokens, budgets[name])
+            if name == "request"
+            else _head_tail(tokens, budgets[name])
```
[the remark this answers is not in the document]
Comment (hasandemirkiran) [reply internal/guard/stepsafety/worker.py]: Fixed in 05c7be3. The Go evaluator rejects fields above 64 KiB or combined input above 128 KiB before backend inference, and the Python worker repeats the guard before tokenizer.encode. Oversized input records the stable input_too_large fail-open category. A timed-out or desynchronized process is now retired and reloaded in the background; the recovery test proves later inference succeeds.

**Comment by hasandemirkiran** [comment]
@greptileai

**Comment by hasandemirkiran** [comment]
@greptileai

**Review by tumberger — APPROVED** [review]
approved for default-off, shadow-only pilot

**Comment by hasandemirkiran** [comment]
@greptileai
